llama_index/chat_engine/cohere_context.py

```diff
@@ -0,0 +1,168 @@
+import asyncio
+from threading import Thread
+from typing import Any, List, Optional, Tuple, Dict
+
+from llama_index.callbacks import trace_method
+from llama_index.chat_engine.types import (
+    CohereAgentChatResponse,
+    CohereStreamingAgentChatResponse,
+    ToolOutput,
+)
+from llama_index.core.base_retriever import BaseRetriever
+from llama_index.core.llms.types import ChatMessage, MessageRole
+from llama_index.llms.llm import LLM
+from llama_index.schema import NodeWithScore
+from llama_index.chat_engine import ContextChatEngine
+from llama_index.llms.cohere_utils import transform_nodes_to_cohere_documents_list
+
+
+class CohereContextChatEngine(ContextChatEngine):
+    """Cohere Context + Citations Chat Engine.
+
+    Uses a retriever to retrieve a context, set the context in the Cohere
+    documents param(https://docs.cohere.com/docs/retrieval-augmented-generation-rag),
+    and then uses an LLM to generate a response.
+
+    NOTE: this is made to be compatible with Cohere's chat model Document mode
+    """
+
+    @trace_method("chat")
+    def chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> CohereAgentChatResponse:
+        if chat_history is not None:
+            self._memory.set(chat_history)
+        self._memory.put(ChatMessage(content=message, role="user"))
+        context_str_template, nodes = self._generate_context(message)
+        prefix_messages = self._get_prefix_messages_with_context(context_str_template)
+
+        all_messages = self._memory.get_all()
+        # transform nodes to cohere documents list
+        documents_list = transform_nodes_to_cohere_documents_list(nodes)
+        # and then uses an LLM to generate a response
+        chat_response = self._llm.chat(all_messages, documents=documents_list)
+        # TODO ask TJ do we need it?
+        ai_message = chat_response.message
+        self._memory.put(ai_message)
+
+        return CohereAgentChatResponse(
+            response=str(chat_response.message.content),
+            citations=chat_response.raw.get("citations", []),
+            documents=chat_response.raw.get("documents", []),
+            sources=[
+                ToolOutput(
+                    tool_name="retriever",
+                    content=str(prefix_messages[0]),
+                    raw_input={"message": message},
+                    raw_output=prefix_messages[0],
+                )
+            ],
+            source_nodes=nodes,
+        )
+
+    @trace_method("chat")
+    def stream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> CohereStreamingAgentChatResponse:
+        if chat_history is not None:
+            self._memory.set(chat_history)
```

Comment:
One thing I would like to clarify is how `CohereContextChatEngine` would manage multiple chats? Currently `_memory` stores a single array of chat messages - Would the application code have to initialize and manage multiple `CohereContextChatEngine` instances?

Reply:
Llama Index provides ChatStores, which we can manage for multiple chats as needed.